File: project-c-sharp/Problem1.cs
namespace advent_of_code_2025;

internal partial class Program
{
    private static void Problem1Part1()
    {
        var fileName = "../../../input/input-1.txt";
        var lines = File.ReadLines(fileName);

        var dialPosition = 50;
        var dialMod = 100;
        var zeroCounter = 0;

        foreach (var line in lines)
        {
            var direction = line[0];
            var distance = int.Parse(line.Substring(1));

            if (direction == 'L')
            {
                dialPosition -= distance;
                while (dialPosition < 0)
                {
                    dialPosition += dialMod;
                }
                dialPosition = dialPosition % dialMod;
            }
            else if (direction == 'R')
            {
                dialPosition += distance;
                dialPosition = dialPosition % dialMod;
            }

            if (dialPosition == 0)
            {
                zeroCounter++;
            }
        }

        Console.WriteLine($"Zero count: {zeroCounter}");
    }

    private static void Problem1Part2()
    {
        var fileName = "../../../input/input-1.txt";
        var lines = File.ReadLines(fileName);

        var dialPosition = 50;
        var dialMod = 100;
        var zeroCounter = 0;

        foreach (var line in lines)
        {
            var prevDialPosition = dialPosition;

            var direction = line[0];
            var distance = int.Parse(line.Substring(1));

            var fullRotations = distance / dialMod;
            var remainderDistance = distance % dialMod;

            if (direction == 'L')
            {
                dialPosition -= remainderDistance;
                if (dialPosition <= 0 && prevDialPosition > 0)
                {
                    zeroCounter++;
                }

            }
            else if (direction == 'R')
            {
                dialPosition += remainderDistance;
                if (dialPosition >= dialMod)
                {
                    zeroCounter++;
                }
            }

            while (dialPosition < 0)
            {
                dialPosition += dialMod;
            }
            dialPosition = dialPosition % dialMod;

            zeroCounter += fullRotations;
        }

        Console.WriteLine($"Zero count: {zeroCounter}");
    }
}

File: project-c-plus-plus/Problem1.cpp
#include <fstream>
#include <iostream>
#include <string>

void Problem1Part1()
{
    std::string file_path = "Input/input-1.txt";
    std::ifstream file (file_path);
    if (!file)
    {
        std::cerr << "Could not open file " << file_path << '\n';
        return;
    }
    
    int dial_position = 50;
    int dial_size = 100;
    int zero_counter = 0;
    
    std::string line;
    while (std::getline(file, line))
    {
        if (line.empty())
        {
            std::cerr << "Found empty line in" << file_path << ", skipping" << '\n';
            continue;
        }
        std::cout << line << '\n';
        
        // First character of the line indicates direction
        char direction = line[0];
        
        // Remaining characters are the rotation distance
        std::string distance_string = line.substr(1);
        int distance = stoi(distance_string);
        
        if (direction == 'L')
        {
            dial_position -= distance;
            while (dial_position < 0)
            {
                dial_position += dial_size;
            }
            dial_position = dial_position % dial_size;
        }
        else if (direction == 'R')
        {
            dial_position += distance;
            dial_position = dial_position % dial_size;
        }
        
        if (dial_position == 0)
        {
            zero_counter++;
        }
    }
    
    std::cout << "Zero count: " << zero_counter << '\n';
}

void Problem1Part2()
{
    std::string file_path = "Input/input-1.txt";
    std::ifstream file (file_path);
    if (!file)
    {
        std::cerr << "Could not open file " << file_path << '\n';
        return;
    }
    
    int dial_position = 50;
    int dial_size = 100;
    int zero_counter = 0;
    
    std::string line;
    while (std::getline(file, line))
    {
        if (line.empty())
        {
            std::cerr << "Found empty line in" << file_path << ", skipping" << '\n';
            continue;
        }
        std::cout << line << '\n';
        
        int previous_position = dial_position;
        
        // First character of the line indicates direction
        char direction = line[0];
        
        // Remaining characters are the rotation distance
        std::string distance_string = line.substr(1);
        int distance = stoi(distance_string);
        
        int full_rotations = distance / dial_size;
        int remainder_distance = distance % dial_size;
        
        if (direction == 'L')
        {
            dial_position -= remainder_distance;
            if (dial_position <= 0 &&
                previous_position > 0)  // Edge case where dial is already at 0 and moves left doesn't count as a click
            {
                zero_counter++;
            }
            
            while (dial_position < 0)
            {
                dial_position += dial_size;
            }
        }
        else if (direction == 'R')
        {
            dial_position += remainder_distance;
            if (dial_position >= dial_size)
            {
                zero_counter++;
            }
        }
        
        dial_position = dial_position % dial_size;
        
        zero_counter += full_rotations;
    }
    
    std::cout << "Zero count: " << zero_counter << '\n';
}